File: test/language/templates/cpp98/StructTemplatedFieldTest.cpp
#include "gtest/gtest.h"

#include "templates/struct_templated_field/StructTemplatedField.h"

namespace templates
{
namespace struct_templated_field
{

TEST(StructTemplatedFieldTest, readWrite)
{
    StructTemplatedField structTemplatedField;
    structTemplatedField.getUint32Field().setValue(42);
    structTemplatedField.getCompoundField().getValue().setValue(42);
    structTemplatedField.getStringField().setValue("string");

    zserio::BitStreamWriter writer;
    structTemplatedField.write(writer);
    size_t bufferSize = 0;
    const uint8_t* buffer = writer.getWriteBuffer(bufferSize);

    zserio::BitStreamReader reader(buffer, bufferSize);
    StructTemplatedField readStructTemplatedField(reader);

    ASSERT_TRUE(structTemplatedField == readStructTemplatedField);
}

} // namespace struct_templated_field
} // namespace templates

File: test/language/templates/java/templates/StructTemplatedFieldTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.struct_templated_field.StructTemplatedField;
import templates.struct_templated_field.Field_uint32;
import templates.struct_templated_field.Field_Compound;
import templates.struct_templated_field.Field_string;
import templates.struct_templated_field.Compound;

public class StructTemplatedFieldTest
{
    @Test
    public void readWrite() throws IOException
    {
        final StructTemplatedField structTemplatedField = new StructTemplatedField();
        structTemplatedField.setUint32Field(new Field_uint32(13));
        structTemplatedField.setCompoundField(new Field_Compound(new Compound(42)));
        structTemplatedField.setStringField(new Field_string("string"));

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        structTemplatedField.write(writer);

        final BitStreamReader reader = new ByteArrayBitStreamReader(
                writer.toByteArray(), writer.getBitPosition());
        final StructTemplatedField readStructTemplatedField = new StructTemplatedField(reader);
        assertTrue(structTemplatedField.equals(readStructTemplatedField));
    }
}
